### PR TITLE
Filter by cellCount to consider summed value (#1862)

### PR DESCRIPTION
**Pull request by danielsotirhos**
Author (before primary review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `HCA` label <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes <sub>or this PR does not need porting</sub> 
- [x] Added `[r]` prefix at start of commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update` <sub>or this PR does not change requirements*.txt</sub>
- [x] Mentioned `make requirements` in UPGRADING.rst <sub>or this PR does not change requirements*.txt</sub>
- [x] Documented upgrade of personal deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `[r]` prefix of commit title
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox` <sub>or commented that sandbox will be skipped</sub>
- [x] Reindexed `sandbox` <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit

Primary reviewer (after pushing merge commit)

- [x] Moved linked issue to Merged
- [ ] Pushed merge commit to Github
- [ ] Moved freebies to Merged column <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Commented on demo expectations <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab
- [ ] Reindexed `dev` <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

### CODE REVIEW

File: test/service/test_response.py [hunk 1]
```diff
@@ -1588,13 +1588,25 @@ def tearDownClass(cls):
...
+                len(hit['cellSuspensions']),
+                sum([cs['totalCells'] for cs in hit['cellSuspensions']])
+            )
+            for hit in response_json['hits']
+        ]
+
     def test_sorting_by_cell_count(self):
         """
-        Verify sorting by 'cellCount' sorts each hit by the sum of 'totalCells'
+        Verify sorting by 'cellCount' considers the sum of 'totalCells'
```
Comment (jessebrennan) [inline test/service/test_response.py:1606]: Nitpick: isn't `totalCells` already a sum? If that's so them I think
```suggestion
        Verify sorting by 'cellCount' considers the value of 'totalCells'
```

**Review by jessebrennan — APPROVED** [review]
This looks good! Nice work. How come you had to add a post-aggregate step? Why can't the cell count be properly calculated during aggregation?

File: test/service/test_response.py (hunk 1 above)
Comment (danielsotirhos) [reply test/service/test_response.py:1606]: > This looks good! Nice work. How come you had to add a post-aggregate step? Why can't the cell count be properly calculated during aggregation?

When the aggregation is performed in `IndexService._aggregate_entity()` it is done per `entity_type`:
```
for entity_type, entities in contents.items():
    ...
    aggregate_contents[entity_type] = entities
```
The post-aggregate step is required to inject the new (and HCA specific) property `total_estimated_cells` at the `entity_type` level (eg. `contents.total_estimated_cells`). Doing it this way allows the `IndexService` to remain generic by putting the HCA specific transformation code in the HCA metadata plugin's transformer.

Comment (danielsotirhos) [reply test/service/test_response.py:1606]: > Nitpick: isn't `totalCells` already a sum? If that's so then I think

It is true that `totalCells` is a summed value in that `cell_suspension` inner entities are grouped by organ, however when filtering and sorting by `totalCells` the value to be considered is the sum of the `totalCells` property across all the groups. I'll modify the docstring to help clarify.

File: test/service/test_response.py [hunk 2]
```diff
@@ -1590,8 +1590,8 @@ def tearDownClass(cls):
 
     def _count_total_cells(self, response_json):
         """
-        Return the number of cell suspension inner entities and total cell count
-        per hit
+        Return per hit the number of cell suspension inner entities and sum
+        value of 'totalCells' from those cell suspension inner entities.
```
Comment (jessebrennan) [inline test/service/test_response.py:1594]: I found the old wording more understandable, but I don't have a strong objection.

**Review by jessebrennan — APPROVED** [review]

File: src/azul/indexer/index_service.py [hunk 3]
```diff
@@ -441,7 +442,7 @@ def _aggregate_entity(self, transformer: Transformer, contributions: List[Contri
             else:
                 aggregator = transformer.get_aggregator(entity_type)
                 if aggregator is not None:
-                    entities = aggregator.aggregate(contents[entity_type])
+                    entities = aggregator.aggregate(entities)
```
Comment (hannes-ucsc) [inline src/azul/indexer/index_service.py:445]: Separate commit like we talked about.

File: src/azul/indexer/transform.py [hunk 4]
```diff
@@ -68,3 +71,12 @@ def get_aggregator(cls, entity_type) -> EntityAggregator:
         entities of types other than X.
         """
         raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def post_aggregate_process(cls, contents: MutableJSON):
```
Comment (hannes-ucsc) [inline src/azul/indexer/transform.py:77]:
```suggestion
    def post_process_aggregate(cls, contents: MutableJSON):
```

Comment (hannes-ucsc) [reply src/azul/indexer/transform.py:77]: type hint return type

File: src/azul/indexer/transform.py [hunk 5]
```diff
@@ -68,3 +71,12 @@ def get_aggregator(cls, entity_type) -> EntityAggregator:
...
+        """
+        Modifies the contents of a document after aggregation and prior to
```
Comment (hannes-ucsc) [inline src/azul/indexer/transform.py:79]:
```suggestion
        Modify the contents of an aggregate document before it is
```

File: src/azul/indexer/transform.py [hunk 6]
```diff
@@ -68,3 +71,12 @@ def get_aggregator(cls, entity_type) -> EntityAggregator:
...
+        being written to Elasticsearch.
```
Comment (hannes-ucsc) [inline src/azul/indexer/transform.py:80]:
```suggestion
        written to the index.
```

File: test/service/test_response.py [hunk 7]
```diff
@@ -1588,13 +1588,25 @@ def tearDownClass(cls):
...
+            )
+            for hit in response_json['hits']
+        ]
+
     def test_sorting_by_cell_count(self):
         """
-        Verify sorting by 'cellCount' sorts each hit by the sum of 'totalCells'
-        when hits contain multiple cell suspension inner entities.
+        Verify sorting by 'cellCount' sorts the hits based on the sum of
+        'totalCells' from the cell suspension inner entities in each hit.
```
Comment (hannes-ucsc) [inline test/service/test_response.py:1607]: Still a little awkward.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/indexer/transform.py [hunk 8]
```diff
@@ -68,3 +71,12 @@ def get_aggregator(cls, entity_type) -> EntityAggregator:
         entities of types other than X.
         """
         raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
```
Comment (hannes-ucsc) [inline src/azul/indexer/transform.py:76]: As I believe we discussed, this doesn't need to be abstract. The default implementation of this method should simply do nothing.

File: src/azul/plugins/metadata/hca/transform.py [hunk 9]
```diff
@@ -139,6 +138,13 @@ def get_aggregator(cls, entity_type):
         else:
             return SimpleAggregator()
 
+    @classmethod
+    def post_process_aggregate(cls, contents: MutableJSON) -> None:
+        cell_count = sum(cs['total_estimated_cells']
+                         for cs in contents['cell_suspensions']
+                         if cs['total_estimated_cells'])
```
Comment (hannes-ucsc) [inline src/azul/plugins/metadata/hca/transform.py:145]:
```suggestion
                         if cs['total_estimated_cells'] is not None)
```

I don't think the optimization of not adding up `0` is not worth the lack of clarity.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/indexer/transform.py [hunk 10]
```diff
@@ -68,3 +71,11 @@ def get_aggregator(cls, entity_type) -> EntityAggregator:
         entities of types other than X.
         """
         raise NotImplementedError
+
+    @classmethod
```
Comment (hannes-ucsc) [inline src/azul/indexer/transform.py:75]: Does not need to be a class method. Remember to only use class methods when it is impossible or prohibitively expensive to obtain an instance.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Could have caught this earlier. No bump.

File: src/azul/indexer/transform.py (hunk 10 above)
Comment (danielsotirhos) [reply src/azul/indexer/transform.py:75]: I believe in this case it needs to be class method as in `IndexService._aggregate()` I only have the transformer class (obtained from `DocumentService.transfomers()`) and no bundle that could be used to create the transformer object.

Comment (hannes-ucsc) [reply src/azul/indexer/transform.py:75]: You are right.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
For demo, attempt to reproduce #1862.

**Pull request merged**